langchain4j-bedrock/src/main/java/dev/langchain4j/model/bedrock/BedrockAnthropicMessageChatModel.java

```diff
@@ -53,47 +69,259 @@ protected Map<String, Object> getRequestParameters(String prompt) {
+            parameters.put("tool_choice", toAnthropicToolChoice(toolChoiceSpecification));
+        }
+
+        if (!toolSpecifications.isEmpty()) {
+            validateModelIdWithToolsSupport();
+            parameters.put("tools", toAnthropicToolSpecifications(toolSpecifications));
+        }
+
         final String body = Json.toJson(parameters);
-        
+
         InvokeModelResponse invokeModelResponse = withRetry(() -> invoke(body), getMaxRetries());
         final String response = invokeModelResponse.body().asUtf8String();
         BedrockAnthropicMessageChatModelResponse result = Json.fromJson(response, getResponseClassType());
-        
-        return new Response<>(new AiMessage(result.getOutputText()),
+
+        return new Response<>(
+            aiMessageFrom(result),
             result.getTokenUsage(),
-            result.getFinishReason());
+            result.getFinishReason()
+        );
     }
-    
+
+    private Object toAnthropicToolChoice(ToolSpecification toolChoiceSpecification) {
+        ObjectNode toolChoiceNode = new ObjectMapper().createObjectNode();
+        toolChoiceNode.put("type", "tool");
+        toolChoiceNode.put("name", toolChoiceSpecification.name());
+        return toolChoiceNode;
+    }
+
+    private void validateModelIdWithToolsSupport() {
+        List<String> anthropicModelIdSplit = Arrays.asList(this.model.split("-"));
+
+        if (anthropicModelIdSplit.size() < 2) {
+            throw new IllegalArgumentException("Tools are currently not supported by this model");
+        }
+
+        String anthropicModelMajorVersion = anthropicModelIdSplit.get(1);
+        if (anthropicModelMajorVersion.contains(":")) {
+            anthropicModelMajorVersion = anthropicModelMajorVersion.split(":")[0];
+        }
+
+        if (!anthropicModelMajorVersion.matches("[0-9]")) {
+            anthropicModelMajorVersion = anthropicModelMajorVersion.replaceAll("[^0-9]", "");
+        }
+
+        if (anthropicModelMajorVersion.isEmpty() || Integer.parseInt(anthropicModelMajorVersion) < 3) {
+            throw new IllegalArgumentException("Tools are currently not supported by this model");
+        }
+    }
+
+    private AiMessage aiMessageFrom(BedrockAnthropicMessageChatModelResponse result) {
+        List<BedrockAnthropicContent> toolUseRequests = result.getContent()
+                .stream()
+                .filter(content -> content.getType().equals("tool_use"))
+                .collect(Collectors.toList());
+
+        if (toolUseRequests.isEmpty()) {
+            return AiMessage.from(result.getOutputText());
+        }
+
+        List<ToolExecutionRequest> toolExecutionRequests = toolUseRequests.stream()
+                .map(toolUseRequest -> ToolExecutionRequest.builder()
+                        .id(toolUseRequest.getId())
+                        .name(toolUseRequest.getName())
+                        .arguments(Json.toJson(toolUseRequest.getInput()))
```

Comment:
Please do not use `Json.toJson()` or `Json.fromJson()`, use `ObjectMapper` instead

Reply:
Done.